add messages regarding random effects grouping factors

FBartos opened the pull request:
fixes 2/3 of: https://github.com/jasp-stats/jasp-test-release/issues/806

1) sets an error on the empty default table if no random effects grouping factor is selected to notify the user about it.
![image](https://user-images.githubusercontent.com/38475991/85052065-9e778600-b198-11ea-8419-c9399461affb.png)

2) adds a note with the random effects grouping factors specification
![image](https://user-images.githubusercontent.com/38475991/85052214-d5e63280-b198-11ea-85a4-a1420cb82e06.png)

FBartos committed:
add messages regarding random effects grouping factors

AlexanderLyNL commented:
Sorry, I should have made this clear before, but please don't put a big red box there, because that feels like a punishment. Instead, show an empty table with that message in the footnote, and the fixed effects labels, but with ... in the output, so it's clear to the user that the analysis is waiting for something.

FBartos committed:
error -> note

FBartos commented:
changing 1) to a note instead
![image](https://user-images.githubusercontent.com/38475991/85057090-8c014a80-b1a0-11ea-89f8-d71607719fa5.png)

AlexanderLyNL commented:
- Yes that looks great. To make it even more obvious, I'd have contGamma on the left, or what ever variables entered in "contGamma" and then dots in the output. 
- The solution with the note is great, the only problem I now have is that it's not that obvious, whereas the previous red box was too obvious
- @TimKDJ what would be good user feedback here? A highlighted "Random effects grouping factors box" <- I noticed something happening with this type of user feedback via the upgrader? Or the note text in green? Or instead of dots in the table the message "Please also specify 'random effects' " but not as a big red box?

FBartos committed:
adds dots to the tables

FBartos committed:
fixes not enabling "Number of trials" for BGLMM

FBartos commented:
So, this would be the output with the dots
![image](https://user-images.githubusercontent.com/38475991/85063464-785ae180-b1aa-11ea-9c95-52da0b439185.png)

And I also fixed when "Number of trials" was not correctly enabled when "Binomial (aggregated)" was selected.

AlexanderLyNL commented:
- I think that a plug-in of "contGamma" under effects (and all other variables names specified under fixed effects variables) could be helpful. 
-  The reason is that it provides the user with feedback acknowledging the user "I understand that you added contGamma" and the dot output conveys that it's waiting for something. @TimKDJ is that still the rule? Or are we going for empty rows now?
- In principle the note already informs the user the reason why it's not producing any results yet
- It would have been easier if random effects were reported, then you can have a random effects table with a column "Random effects" and under it ... stating that the analysis needs a random effect. 
- The note takes over the role of the previous point.
- Again, we need something more prominent than the note, especially for those with old eyes. But that's a different issue. 

An alternative approach would be to change the UI:
- enter all the variables into "Covariates" (instead of fixed effects): Produce standard GLM result
- Section called "Random effects" takes in subset of variables which can be selected and specified as random effects
- The idea is loosely based on the UI of linear regression, where we specify nuisance parameters under the "Model" section

EJWagenmakers commented:
Without having given this specific suggestion much thought (Frantisek, Lexi, let me know if you want to discuss this in a zoom meeting or something!) I think we should at least give Frantisek/Henrik's GUI a go for 0.13 -- we are in the final, final stages of getting out a new version and unless there is a blatant error I think we should avoid making structural changes. We can always change the GUI for the next version.

TimKDJ commented:
I think for now a footnote should suffice. But maybe it should read `This analysis requires at least one random effects grouping factor to run`. At that point everyone should be able to work out why the analysis is in a waiting state.

The long term solution is to make the analysis work without the grouping factor as described in [this issue](https://github.com/jasp-stats/INTERNAL-jasp/issues/932); it's obviously better to make the analysis do what a user would reasonably expect it to :)

FBartos committed:
missing RE & aggreagation note + fixing binomial aggregated BGLMM

FBartos commented:
So, I changed the message as suggested (it's shown only if the Random-effects grouping factor is missing):
![image](https://user-images.githubusercontent.com/38475991/85140527-a80af780-b245-11ea-8c2e-215fbad6e460.png)

and added a message to address this issue (https://github.com/jasp-stats/jasp-test-release/issues/836)- it's not clear that one needs to specify the "Number of trials" variable in order to run the analysis with "Binomial (aggregated)" family 
![image](https://user-images.githubusercontent.com/38475991/85140797-12239c80-b246-11ea-89ae-dc40a95bb304.png)

I also fixed one issue that would prevent the BGLMM with "Binomial (aggregated)" family to run

TimKDJ commented on JASP-Engine/JASP/R/MixedModelsCommon.R:
```diff
@@ -552,7 +551,17 @@
+        ANOVAsummary$addFootnote(.mmMessageMissingRE)
+      }
+      if (type == "GLMM") {
+        if (options$family == "binomial_agg" &
```
`&&`

TimKDJ commented on JASP-Engine/JASP/R/MixedModelsCommon.R:
```diff
@@ -2904,6 +2927,18 @@
   gettextf("Type %s Sum of Squares",type)
 }
+.mmMessageREgrouping    <- function(RE_grouping_factors) {
+  gettextf(
+    "The following %s used %s random effects grouping %s: %s.",
+    ifelse(length(RE_grouping_factors) < 2, "variable is", "variables are"),
+    ifelse(length(RE_grouping_factors) < 2, "as a", "as"),
+    ifelse(length(RE_grouping_factors) < 2, "factor", "factors"),
+    ifelse(length(RE_grouping_factors) == 0, "...", paste0("'", RE_grouping_factors, "'", collapse = ", "))
+  )
+}
```
Use `sprintf(ngettext())` for this. Aside from the strings not being in `gettext()` the structure of this sentence might need to be entirely different in another language.

FBartos replied:
Yeah, thanks. I keep forgetting those.

What do you advise to do about it then?

FBartos replied:
I could maybe create two complete sentences for each case

TimKDJ replied:
Something like:
```
sprintf(ngettext(length(RE_grouping_factors) < 2,
                 "The following variable is used as random effects grouping factor: %s.",
                 "The following variables are used as random effects grouping factors: %s."),
        ifelse(length(RE_grouping_factors) == 0, "...", paste0("'", RE_grouping_factors, "'", collapse = ", ")))
```

untested though

FBartos committed:
changing mesages for translation

FBartos committed:
if(){}else{} to ngettext

TimKDJ commented on JASP-Engine/JASP/R/MixedModelsCommon.R:
```diff
@@ -2986,32 +2968,25 @@
-  }
+  sprintf(
+    ngettext(
+      value < 2,
```
Unless you specifically want to group `0` and `1` together you should use `ngettext(length(value), "..", "..")` (so no boolean, but an integer as argument to `ngettext()`)

`0 observation was removed due to missing values.` is not correct.

This goes for quite a few of these changed lines

FBartos replied:
the 's' is actually correct at the end - there might be more missing values within one observation. We do not check those.

These messages aren't called when the value == 0

TimKDJ replied:
We've solved this over Slack, but:

`0 observation was removed` is correct not though :)
These are your sentences:
```
 "%i observation was removed due to missing values.",
 "%i observations were removed due to missing values."
```

FBartos replied:
0 observation was removed cannot happen. The function is called only if `value > 0`

TimKDJ replied:
I know, but the English is not correct. It was in response to this:

> the 's' is actually correct at the end - there might be more missing values within one observation. We do not check those.

FBartos replied:
oh, my English is probably not good enough to understand the problem with my English :D

FBartos committed:
integers for ngettext
